src/1-theory/3-input-and-output.adoc

```diff
@@ -0,0 +1,593 @@
+
+Finally, do not try to be clever by mixing JSON with form data:
+
+.This is nonsense. Do not do it.
+[source,http]
+----
+POST /checkins HTTP/1.1
+Host: api.example.org
+Content-Type: application/x-www-form-urlencoded
+
+json="{
+    \"checkin\": {
+        \"place_id\" : 1,
+        \"message\": \"This is a bunch of text.\",
+        \"with_friends\": [1, 2, 3, 4, 5]
+    }
+}"
+----
+
+This actually happens surprisingly often in the wild.
+
+image::images/input-output-json-json.jpg[]
+
+Some folks even
+
+=== Why many prefer JSON to XML
+
+Any modern API you interact with will support JSON, or there is some fancy
+binary format being used. Sometimes APIs will support XML too, especially if the
+API is maintained by an older financial services company. XML generally takes
+more memory to convey the same amount of data as JSON, as its similarity to HTML
+means it requires both a start and an end tag containing the same name.
+
+Beyond purely the size of the data being stored, XML is rather terrible with
+handling different data types. That might not worry dynamic language developers
+all that much, but look at this:
+
+.An example of a bunch of different data types in JSON.
+[source,javascript]
+----
+{
+  "place": {
+    "id" : 1,
+    "name": "This is a bunch of text.",
+    "is_true": false,
+    "maybe": null,
+    "empty_string": ""
+  }
+}
+----
+
+.The same example but in XML.
+[source,xml]
+----
+<places>
+    <place>
+        <id>1</id>,
+        <name>This is a bunch of text.</name>
+        <is_true>0</is_true>
+        <maybe />
+        <empty_string />
+    </place>
+</places>
+----
+
+Basically, in XML, _everything_ is considered a string, meaning integers,
+booleans, and nulls can be confused. Both `maybe` and `empty_string` have the
+same value, because there is no way to denote a null value either. Gross.
+
+Work out which content type(s) you actually need, and _stick to that_. The
+Flickr API used to support `lolcat` as a joke, and that was probably the result
+of a hack project in which the development team were only paid with cold pizza.
+JSON is fine.
+
+CSV can be pretty handy as an export format too, especially if your API is
+offering data for any sort of "Reports".
+
+Whatever you decide to offer, make it very clear on your documentation what
+formats are supported. HTTP clients can request a specific `Content-Type` be
+used for the response by placing it in the `Accept` header on the request, and
+if that type is not available you can return a 406 Not Acceptable response,
+which is one of many errors an API can return when a request goes wrong.
+
+image::images/406.jpg[406 - Not Acceptable - Kitty does not want a bath.]
+
+// TODO when to return a body
+// - PATCH/PUT/POST returning content, maaaaybe it should be Content-Location
+// instead of a direct return
+// What (if anything) do you return in the response body for POST, PUT and PATCH?
```

Comment:
Also worth clarifying what to return when the resource exists but it's empty? Eg. a list of products that has no products in it at the mo. I've seen some people return 404 when I think it should have been 200 with an empty body, no? I think it's a common mistake / source of confusion.